panther_cli/main.py

```diff
@@ -170,6 +172,59 @@ def zip_policies(args: argparse.Namespace) -> Tuple[int, str]:
         'zip', args.policies)
 
 
+def upload_policies(args: argparse.Namespace) -> Tuple[int, str]:
+    """Tests, validates, packages, and then uploads all policies into a Panther deployment.
+
+    Returns 1 if the policy tests, validation, or packaging fails.
+
+    Args:
+        args: The populated Argparse namespace with parsed command-line arguments.
+
+    Returns:
+        A tuple of return code and the archive filename.
+    """
+    return_code, archive = zip_policies(args)
+    if return_code == 1:
+        return return_code, ''
+
+    client = boto3.client('lambda')
+
+    with open(archive, 'rb') as analysis_zip:
+        zip_bytes = analysis_zip.read()
+        payload = {
+            'resource':
+                '/upload',
+            'HTTPMethod':
+                'POST',
+            'Body':
+                json.dumps({
+                    'Data': base64.b64encode(zip_bytes).decode('utf-8'),
+                    'UserID': 'c273fd96-88d0-41c4-a74e-941e17832915',
+                }),
+        }
+
+        logging.info('Uploading pack to Panther')
+        response = client.invoke(FunctionName='panther-analysis-api',
```

Comment:
Interesting, so you're invoking the Lambda function directly instead of going through API gateway? It certainly is easier to implement but the idea is that all invocations of the lambda function are funneled through the gateway so we can get logging + metrics + WAF integration and all the other API gateway features.

@kostaspap was able to talk to API gateway from the Python analysis engine. If it's relatively easy, I think I'd rather invoke the API gateway directly to be consistent. The Lambda input/output is an internal detail subject to change whereas the API gateway is a publicly broadcast model that all services rely on.

I could go either way on this, curious what @kostaspap thinks?

Reply:
I guess I've never used the API gateway logs, I've just been using the lambda functions CloudWatch logs which are pretty helpful and those work with this method.

My main concern with going through API gateway is figuring out the URL since we're running the CLI with no awareness of their deployment, with invoking the lambda function we just give the name but there's no way to just give the name for API gateway as far as I'm aware. I really don't want to make users paste in the random generated URL. I guess we can list all the API gateways, find the right one, then get the URL?

I didn't realize we were avoiding invoking the Lambda functions directly, I can look into using API gateway again tomorrow.

Reply:
> My main concern with going through API gateway is figuring out the URL

Hmm fair point. You can just describe the `panther-app` CloudFormation stack to get the API ID, but it certainly seems like a lot more effort.

I'm fine with leaving this as-is, we can reconsider when we standardize / publish a unified API

Reply:
GatewayAPI or Lambda are both "back doors" to our system. From any user's perspective the Panther API is actually unified, it is GraphQL API (AppSync). It might be that our frontend team is currently working on it it, but technically AppSync is backend :)
We should be going through that endpoint, maybe through the ALB (since that's what users see) but not through GatewayAPI or Lambda. 
ALB would actually be preferable, so if anyone puts a CNAME to our ALB, they can use that for uploading rules. 

Is there a way to use Cognito to authenticate? Requiring AWS credentials to bulk upload rules is really going through the backdoor. The mechanism to do that should use the same stack as the equivalent operation from the UI, maybe using a different auth mechanism (e.g. some access token). Bulk uploads need to go through RBAC and need to be audited like any other operation :)

I get it that it's early days and we want to provide an easy way for users to upload rules, but this is something we should fix eventually. We will need that so people can code manage and deploy rules through some CI/CD pipeline.

Reply:
I agree we should go through the front door eventually.

Currently, I don't think there is a way for users to authenticate without copying in the randomly generated URL, the username, password, and MFA code. Which is fine, although maybe we could add support for a config file that stores the URL and username at least.

For some reason, I thought we were discussing moving away from AppSync though? I didn't want to add a workflow based on features that are going away.